streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -1938,115 +1938,134 @@ object ZChannel {
-                               val latch    = Promise.unsafe.make[Nothing, Unit](fiberId)(Unsafe.unsafe)
-                               for {
-                                 size <- cancelers.size
-                                 _    <- ZIO.when(size >= n)(cancelers.take.flatMap(_.succeed(())))
-                                 _    <- cancelers.offer(canceler)
-                                 raceIOs =
-                                   ZIO.scopedWith { scope =>
-                                     (incoming >>> channel)
-                                       .toPullInAlt(scope)
-                                       .flatMap(evaluatePull(_).race(canceler.await.interruptible))
-                                   }
-                                 _ <- permits
-                                        .withPermit(latch.succeed(()) *> raceIOs)
-                                        .interruptible
-                                        .forkIn(childScope)
-                                 _ <- latch.await
-                               } yield ()
-                             }
-                       }
-        _ <- pullStrategy(fiberId).forever
-               .catchAllCause(cause =>
-                 cause.failureOrCause match {
-                   case Left(_: Left[OutErr, OutDone]) => outgoing.offer(Exit.failCause(cause))
-                   case Left(x: Right[OutErr, OutDone]) =>
-                     permits.withPermits(n.toLong)(ZIO.unit).interruptible *>
-                       lastDone.get.flatMap {
-                         case Some(lastDone) => outgoing.offer(Exit.fail(Right(f(lastDone, x.value))))
-                         case None           => outgoing.offer(Exit.fail(Right(x.value)))
-                       }
-                   case Right(cause) => outgoing.offer(Exit.failCause(cause.map(Left(_))))
-                 }
-               )
-               .raceFirst(awaitErrorSignal(childScope, fiberId)(errorSignal))
-               .forkIn(scope)
-      } yield {
-        lazy val consumer: ZChannel[Env, Any, Any, Any, OutErr, OutElem, OutDone] =
-          unwrap[Env, Any, Any, Any, OutErr, OutElem, OutDone] {
-            outgoing.take.flatten.foldCause(
-              cause =>
+      ZIO.fiberIdWith { fiberId =>
+        ZIO.suspendSucceed {
+          val input          = SingleProducerAsyncInput.unsafe.make[InErr, InElem, InDone](fiberId)
+          val incoming       = ZChannel.fromInput(input)
+          val n0             = n.toLong
+          val bufferSize0    = bufferSize
+          val mergeStrategy0 = mergeStrategy
+          val outgoing       = Queue.unsafe.bounded[Result](bufferSize0, fiberId)(Unsafe.unsafe)
+          val cancelers      = Queue.unsafe.unbounded[Promise[Nothing, Unit]](fiberId)(Unsafe.unsafe)
+          val lastDone       = Ref.unsafe.make[OutDone](null.asInstanceOf[OutDone])(Unsafe.unsafe)
```

Comment:
I'm really reluctant on passing `null` to anything that can end up in a ZIO effect containing a `null` value. I think it's better to use `AtomicReference(null)` directly

Reply:
Done. Please have a look. I made the choice to wrap the usage of the `AtomicReference` in `ZIO.succeed` and not in `Exit.succeed` to avoid executing the wrapped code at the wrong moment in time. I can be wrong tho 🤔

Reply:
Might be unnecessary indeed but better be safe than sorry on that front